### PR TITLE
Range-ify std.path.setExtension by adding setExt()

### PR DESCRIPTION
**Pull request by WalterBright**
http://dlang.org/phobos/std_path.html#.setExtension

100% unit test coverage

This is a reboot of https://github.com/D-Programming-Language/phobos/pull/2423 which was itself a reboot of https://github.com/D-Programming-Language/phobos/pull/2149

The recent addition of std.utf.byUTF makes this function much more convenient to write.

### CODE REVIEW

File: std/path.d [hunk 1]
```diff
@@ -986,15 +986,23 @@ unittest
...
     if (isSomeChar!C1 && !is(C1 == immutable) && is(Unqual!C1 == Unqual!C2))
 {
-    if (ext.length > 0 && ext[0] != '.')
-        return cast(typeof(return))(stripExtension(path)~'.'~ext);
-    else
-        return cast(typeof(return))(stripExtension(path)~ext);
+    import std.array;
+    try
+    {
+        return cast(typeof(return))setExt(path, ext).array;
```
Comment (andralex) [inline std/path.d]:
```d
return setExt(path, ext).to!(typeof(return));
```

File: std/path.d [hunk 2]
```diff
@@ -1053,6 +1052,47 @@ unittest
...
+         is(StringTypeOf!R)) &&
+        isSomeChar!C)
+{
+    import std.range : only, chain;
+    import std.utf : byUTF;
+
+    alias CR = Unqual!(ElementEncodingType!R);
+    if (ext.length > 0 && ext[0] == '.')
+        ext = ext[1 .. $];              // remove any leading . from ext[]
+    auto dot = only(cast(CR)'.');
```
Comment (andralex) [inline std/path.d]: `to`

File: std/path.d [hunk 3]
```diff
@@ -1053,6 +1052,47 @@ unittest
...
+    if (ext.length == 0)
+        dot.popFront();                 // so dot is an empty range, too
+    return chain(stripExtension(path).byUTF!CR, dot, ext.byUTF!CR);
```
Comment (andralex) [inline std/path.d]: Simpler logic: don't pop from ext, and insert an optional dot.

```d
auto dot = only(CR('.'));
if (ext.startsWith('.'))
{
    dot.popFront();
}
return chain(stripExtension(path).byUTF!CR, dot, ext.byUTF!CR);
```

Comment (WalterBright) [inline std/path.d]: That isn't quite the same logic. If ext is empty, the . should not be included. Your rewrite includes it.

File: std/path.d (hunk 2 above)
Comment (WalterBright) [inline std/path.d]: why?

File: std/path.d (hunk 1 above)
Comment (WalterBright) [inline std/path.d]: what's wrong with cast?

**Comment by andralex** [comment]
nice

**Comment by andralex** [comment]
Auto-merge toggled on

**Comment by CyberShadow** [comment]
`setExt` and `setExtension`?

Seriously?

How is anyone expected to remember which does which?

As if `std.path` wasn't confusing enough with the deprecated functions.

**Comment by CyberShadow** [comment]
Looks like I'm not the only one who thinks this way:

https://github.com/D-Programming-Language/phobos/pull/2149#issuecomment-42867964

@WalterBright Your other proposal, `withExtension`, is much better. Please rename it before the next release.

**Comment by CyberShadow** [comment]
> Please rename it before the next release.

Looks like I'll have to do it myself.

https://github.com/D-Programming-Language/phobos/pull/3426

**Pull request merged**